src/dashboard-refactor/logic.ts

```diff
@@ -0,0 +1,720 @@
+
+type EventHandler<EventName extends keyof Events> = UIEventHandler<
+    State,
+    Events,
+    EventName
+>
+
+export class DashboardLogic extends UILogic<State, Events> {
+    private searchBG = runInBackground<SearchInterface>()
+
+    constructor(private options: DashboardDependencies) {
+        super()
+    }
+
+    getInitialState(): State {
+        return {
+            searchResults: {
+                results: {},
+                pageData: {
+                    allIds: [],
+                    byId: {},
+                },
+                noteData: {
+                    allIds: [],
+                    byId: {},
+                },
+                searchType: 'pages',
+                areAllNotesShown: false,
+                searchState: 'pristine',
+                paginationState: 'pristine',
+            },
+            searchFilters: {
+                searchQuery: '',
+                isSearchBarFocused: false,
+                domainsExcluded: [],
+                domainsIncluded: [],
+                isDateFilterActive: false,
+                isDomainFilterActive: false,
+                isTagFilterActive: false,
+                searchFiltersOpen: false,
+                tagsExcluded: [],
+                tagsIncluded: [],
+                dateFromInput: '',
+                dateToInput: '',
+            },
+        }
+    }
+
+    /* START - Misc event handlers */
+    searchPages: EventHandler<'searchPages'> = async ({
+        previousState: { searchFilters },
+    }) => {
+        const emit = (state: TaskState) =>
+            this.emitMutation({
+                searchResults: { searchState: { $set: state } },
+            })
+
+        try {
+            emit('running')
+            const result = await this.searchBG.searchPages({
+                contentTypes: { pages: true, highlights: false, notes: false },
+                endDate: searchFilters.dateTo,
+                startDate: searchFilters.dateFrom,
+                query: searchFilters.searchQuery,
+                domainsInc: searchFilters.domainsIncluded,
+                domainsExc: searchFilters.domainsExcluded,
+                tagsInc: searchFilters.tagsIncluded,
+                tagsExc: searchFilters.tagsExcluded,
+            })
+            emit('success')
+
+            const {
+                noteData,
+                pageData,
+                results,
+            } = utils.pageSearchResultToState(result)
+
+            this.emitMutation({
+                searchResults: {
+                    results: { $set: results },
+                    pageData: { $set: pageData },
+                    noteData: { $set: noteData },
+                },
+            })
+        } catch (err) {
+            emit('error')
+        }
+    }
+
+    searchNotes: EventHandler<'searchNotes'> = async ({
+        previousState: { searchFilters },
+    }) => {
+        const emit = (state: TaskState) =>
```

Comment:
Also put this in the Documentation inbox:
https://www.notion.so/worldbrain/executeUITask-utility-UILogic-function-in-Memex-Social-0e2e0c8857294ad9b98ae598c9d3b8a4

Reply:
Uh-huh! I am familiar with the said util fn, though I ran into a problem with its use here. Which is, the current version in Memex ext only supports top-level state keys. I googled a bit into how I could extend the type system to support arbitrarily nested keys, and after a bit of reading concluded that this is not something I wanted to be spending time on right now, hence opting for skipping it and manually mutating that UITaskState for now.

Though it looks like the Memex Social version of that function affords this already! I will update this in Memex ext. Though I'm feeling this should live somewhere shared, else we'll run into the same situation of one version lagging behind the other in the future. What's the reason you don't put this in the `ui-logic-core` package? It seems like something to me that isn't directly coupled to Memex ext or any of our other codebases

Reply:
I want to keep `ui-logic-core` minimal for now, because I feel its evolution is not finished yet. For example the pattern for nested stuff is something I only implemented experimentally last sprint. By doing the small scale experiments I can still correct course if something doesn't quite work out (like I did already with the return value.) Before even sharing it with the extension, it's good to know what might break.